engine/execution/computation/manager.go

```diff
@@ -67,24 +69,40 @@ func New(
 		return nil, fmt.Errorf("cannot create block computer: %w", err)
 	}
 
+	programsCache, err := NewProgramsCache(programsCacheSize)
+	if err != nil {
+		return nil, fmt.Errorf("cannot create programs cache: %w", err)
+	}
+
 	e := Manager{
 		log:           log,
 		me:            me,
 		protoState:    protoState,
 		vm:            vm,
 		vmCtx:         vmCtx,
 		blockComputer: blockComputer,
+		programsCache: programsCache,
 	}
 
 	return &e, nil
 }
 
+func (e *Manager) getProgramsOrEmpty(blockID flow.Identifier) *fvm.Programs {
+	programs := e.programsCache.Get(blockID)
+	if programs == nil {
+		return fvm.NewEmptyPrograms()
+	}
+	return programs
+}
+
 func (e *Manager) ExecuteScript(code []byte, arguments [][]byte, blockHeader *flow.Header, view *delta.View) ([]byte, error) {
 	blockCtx := fvm.NewContextFromParent(e.vmCtx, fvm.WithBlockHeader(blockHeader))
 
 	script := fvm.Script(code).WithArguments(arguments...)
 
-	err := e.vm.Run(blockCtx, script, view)
+	programs := e.getProgramsOrEmpty(blockHeader.ID())
```

Comment:
this won't be thread safe right? sharing programs between scripts and transactions (it might be safer to ignore the cache for scripts for now)

Reply:
`Programs` itself is thread-safe. However now I realized the script has a potential of contaminating cache for the blocks. Scripts should be executed with a copy/child of given blocks Programs, which will make it thread-safe even if the Programs wouldn't